refactor(vcs): VCS configuration to support common and custom attributes

wkl3nk opened the pull request:
While VCS implementations are already plugins, they are not yet configurable. VCS implementations require common configurations (e.g., `revision`, `recursive`) and should support also VCS-specific configurations **if they are consumed via their API**. 

This allows to add functionality to individual VCS implementations without the need to implement them for all of them.

This refactoring keeps the **common** configuration attributes as they are,  while VCS-specific configurations are stored generically in an `options` attribute. The effects are limited to usage through the command-line interface of the
_Downloader_ tool, designed primarily for scripted operations. In other stages of the ORT pipeline these  configuration options have no effect.

The change also adds VCS-specific configuration options for Git: The strategy for checking out repositories with submodules now allows to only check out **the first layer** of submodules.

Example configuration in `config.yml`:
```
ort:
  analyzer:
    versionControlSystems:
      Git:
        options:
          submoduleHistoryDepth: 2
          updateNestedSubmodules: false

```

The change is split up into 2 commits:
* First one deals with refactoring to use a new plugin interface for version control systems, that also allows to supply specific configurations.
* Second one is for 2 configuration options for Github VCS, dealing with submodules.

For more details, please have a look at the respective commits.

Fixes #8556.

fviernau commented:
I believe the options for configuring submodule recursion could be injected via package curations. Have you considered this?

wkl3nk commented:
@fviernau Hello, can you explain a little bit more please?

fviernau commented:
> @fviernau Hello, can you explain a little bit more please?

I believe it makes sense to have the ability to configure this separate per dependency.
I expect the following trade-off to be reasonable:
1. Enable submodule recursion by default
2. Disable submodule recursion per package curation (after reviewing that this is ok, scan won't miss anything relevant)

sschuberth commented on downloader/src/main/kotlin/VersionControlSystem.kt:
```diff
@@ -387,7 +388,8 @@ abstract class VersionControlSystem(
         path: String = "",
-        recursive: Boolean = false
+        recursive: Boolean = false,
+        options: Options = emptyMap() // VCS-specific options only
```
Just adding an unused option to the function signature does not fix anything, so you should not say so in the commit message. Most importantly, these options are not read from the configuration file and passed on.

wkl3nk replied:
I hope I have improved the commit message now, also emphasising that this options are only available if the VCS plugin is consumed using the API.

sschuberth replied:
> options are only available if the VCS plugin is consumed using the API.

I fear that's not enough. For the PR to be acceptable, VCS plugins should be configurable via `config.yml` entries, [just like package managers like *GradleInspector*](https://github.com/oss-review-toolkit/ort/blob/6bb98df0cc4071155d333a9296cec6d6ef52c4e6/model/src/main/resources/reference.yml#L107-L111), independently of any programmatic use / ORT Server context.

wkl3nk replied:
I will make myself familiar with it.

wkl3nk replied:
Now the VCS-specific configuration can also be modified via `config.yml`, I added an example in `reference.yml` (see [here](https://github.com/boschglobal/oss-review-toolkit/blob/a7c9533513dc8ea798af99f0bbe83da9a9eace46/model/src/main/resources/reference.yml#L170C1-L179C39)). 

It is also possible now to use the _ort download_ command on a problematic repository with a myriad of nested submodules and it was actually possible to reduce the submodule-activities to only the first layer of submodules.

sschuberth commented on plugins/version-control-systems/git/src/main/kotlin/Git.kt:
```diff
@@ -60,14 +63,50 @@ import org.ossreviewtoolkit.utils.ort.showStackTrace
     "git://" to "https://"
 )
 
+enum class OptionKey(val key: String, val defaultValue: String, val deprecated: Boolean = false) {
```
All other plugins that take options just use constants for key names, and I believe we should do the same here, for consistency and to reduce complexity. I guess this will get reworked anyway once VCS plugins get migrated to @mnonnenmacher's new plugin API.

wkl3nk replied:
Just saw it for the Gradle and GradleInspector plugins. Makes sense to do it in the same way.

sschuberth commented on plugins/version-control-systems/git/src/main/kotlin/Git.kt:
```diff
@@ -60,14 +63,50 @@ import org.ossreviewtoolkit.utils.ort.showStackTrace
+    // submodules on the top-level should be initialized, updated, and cloned.
+    UPDATE_NESTED_SUBMODULES("updateNestedSubmodules", "true"),
+
+    // Example for deprecating a configuration option
+    DO_NOT_USE("doNotUse", "some-value", deprecated = true);
```
We should not keep code like this.

sschuberth commented on plugins/version-control-systems/git/src/main/kotlin/Git.kt:
```diff
@@ -178,9 +217,24 @@ class Git : VersionControlSystem(GitCommand) {
 
-                updateWorkingTreeWithoutSubmodules(workingTree, git, revision).mapCatching {
+                val optionsValidationResult = OptionKey.validate(options)
+                optionsValidationResult.warnings.forEach { logger.warn { it } }
+                optionsValidationResult.warnings.forEach { logger.error { it } }
```
It does not make sense to log these as both warnings and errors.

fviernau commented:
> I believe it makes sense to have the ability to configure this separate per dependency. I expect the following trade-off / approach to be reasonable for some users:
> 
>     1. Enable submodule recursion by default
> 
>     2. Disable submodule recursion per package curation (after reviewing that this is ok, scan won't miss anything relevant)

To add to this, I believe in order to implement the above it would be necessary to pass
the config flags as parameter to the functions of the VCS implementations which make the cloning.

We could right away unify this instead of introducing two ways to tell the behavior. In particular,
even though Git, GitRepo and so forth are plugins, I believe it would be fine to assume that these are built in.
So, we can add the configuration to `OrtConfiguration` or any model under it, instead of to a plugin configuration.
Then, the caller could just pass whatever is configured to the VCS instance as param. 
Any thoughts on that?

wkl3nk commented:
@fviernau I think I understand your concerns. However, my key issue is that there are very large repositories that act as kind of umbrella around other repositories, having already **more than 50 git submodules in the first layer**. These submodules then also have submodules and so on. One can't even be sure that there are no cycles in this nested submodules. **ORT currently cannot even download these repositories**, we tried it out and stopped it after several hours. 
On the other hand, the project developers say **that they usually only check out the submodules of the first layer**, nothing more. This change request is about just that: Enable ORT to checkout out just the first layer of submodules and not the complete nested, possibly never ending tree of nested submodules.

fviernau commented:
> @fviernau I think I understand your concerns. However, my key issue is that there are very large repositories that act as kind of umbrella around other repositories, having already **more than 50 git submodules in the first layer**. These submodules then also have submodules and so on. One can't even be sure that there are no cycles in this nested submodules. **ORT currently cannot even download these repositories**, we tried it out and stopped it after several hours.
> On the other hand, the project developers say **that they usually only check out the submodules of the first layer**, nothing more. This change request is about just that: Enable ORT to checkout out just the first layer of submodules and not the complete nested, possibly never ending tree of nested submodules.

My concerns were not about the _what_ but about the _how_. So, I generally agree that it makes sense that ORT provides a way to configure this. It also wasn't meant as a strict concern, but I wanted to make sure this is considered before committing to some solution.

wkl3nk commented:
@fviernau wrote
> can you confirm that the issue happens during provenance resolution?

Would "provenance resolution" mean that the Analyzer has scanned for dependency management files? If yes, then the answer is: No - The inflexibility lies in the downloader, it is the default to do a "_git submodule --init --recursive"_ to all submodules. So the Analyzer never ever kicks in, because it just is not possible in a reasonable time and space to download all these > 50 nested submodules and their respective submodules and so on. That's the reason for going in the direction to be able to configure the ORT Downloader in `config.yml`.

fviernau commented:
> Would "provenance resolution" mean that the Analyzer has scanned for dependency management files? 

Oh, I see. It's with the `project` sources, not with any of the `Package`s sources aka. dependencies.

wkl3nk commented:
Hello @sschuberth , the UI tells me _1 change requested_ from your side, but unfortunately I cannot find it. Maybe it was overwritten somehow when I did a _force push_? Could you please support me in locating this change request. 
Besides that, Martin is back from vacation, if there is need to discuss this.

sschuberth commented on downloader/src/main/kotlin/Downloader.kt:
```diff
@@ -270,8 +270,13 @@ class Downloader(private val config: DownloaderConfiguration) {
             return RepositoryProvenance(pkg.vcsProcessed, pkg.vcsProcessed.revision)
         }
 
+        val vcsConfig = config.getVersionControlSystemConfiguration(applicableVcs.type)
```
Commit message:

> (e.g., `revision`, `recursive`)

`revision` probably is not such a good example as it usually changes on a call-by-call basis. Better use `priority` as another example, as that can be static per ORT run.

wkl3nk replied:
Agreed. I convert the PR back to draft and create a solution based on TypedConfigurablePluginFactory, like e.g. ScannerWrapperFactory does.

sschuberth commented on downloader/src/main/kotlin/Downloader.kt:
```diff
@@ -270,8 +270,13 @@ class Downloader(private val config: DownloaderConfiguration) {
         }
 
+        val vcsConfig = config.getVersionControlSystemConfiguration(applicableVcs.type)
+        if (vcsConfig != null) {
```
Better use `?.also { ... }` on the previous line.

sschuberth commented on downloader/src/main/kotlin/VersionControlSystem.kt:
```diff
@@ -380,14 +383,17 @@ abstract class VersionControlSystem(
         path: String = "",
-        recursive: Boolean = false
+        recursive: Boolean = false,
+        vcsConfiguration: VersionControlSystemConfiguration? = null
```
Conceptually, this `vcsConfiguration` should not be passed to a specific function, but to the constructor. IMO the right way to do would be to introduce a `VersionControlSystemFactory` that inherits from `TypedConfigurablePluginFactory`, like e.g. `ScannerWrapperFactory` does.

wkl3nk replied:
Agreed. I convert the PR back to draft and create a solution based on TypedConfigurablePluginFactory, like e.g. ScannerWrapperFactory does.

sschuberth reviewed (commented):
I believe this PR needs some refactoring to be in line with the plugin factory concept of other plugins, and to not interfere with the new plugin API created by @mnonnenmacher.

wkl3nk commented:
Now using the new _Plugin_ concept that is based on `TypedConfigurablePluginFactory` and is already was implemented for the scanners before. VCS classes now have a factory inside, and can have a specific configuration class, that is injected in the contructor of the VCS implementation by the factory. As an example, I added such a specific configuration class for Git.

sschuberth commented:
> Now using the new _Plugin_ concept that is based on `TypedConfigurablePluginFactory`

Before I start review, I'd like @mnonnenmacher to comment whether this "intermediate step" is meaningful to do (now that it's already there...) instead of straight migrating all VCS to the KSP-based plugin mechanism in the context of https://github.com/oss-review-toolkit/ort/issues/9403.

mnonnenmacher commented:
> > Now using the new _Plugin_ concept that is based on `TypedConfigurablePluginFactory`
> 
> Before I start review, I'd like @mnonnenmacher to comment whether this "intermediate step" is meaningful to do (now that it's already there...) instead of straight migrating all VCS to the KSP-based plugin mechanism in the context of #9403.

It would have made sense to use the KSP plugin API right away, however, now that it is implemented, the effort to do it in this PR or another one would be the same, so both would be fine with me.

@wkl3nk However, I have a conceptual question: In a usual ORT pipeline the downloader is used in two places:
* Before running the analyzer to download the project repository.
* By the scanner to download the sources of the dependencies.
The new options you introduce are to my understanding only relevant for the first step to limit the downloaded sources of the project itself. But they should not be applied within the scanner, because you cannot know beforehand if the dependencies use any submodules and if they need to be scanned or not.
Did you consider this during the implementation?

wkl3nk commented:
> Where can this functionality currently be used?

@mnonnenmacher The functionality can out of the box now be used for the _Downloader_ stage, meaning in the typical workflow where the ORT CLI  is first used to download the repository. It supports adding VCS specific configurations for e.g. Git. For instance, you can configure in _.ort/config/config.yml_ if only the first layer of submodules shall be downloaded, in case you need to deal with repositories with many many nested submodule layers. Once the repository is downloaded, it can be analyzed and scanned with other options of the ORT CLI.

In ORT Server, where we don't have this _Downloader_ stage, and instead the _Downloader_ is **instantiated and used in the code**, just before the _Analyzer_, this change also allows to configure the VCSs in a flexible way. How this is done would be a question within the context of the ORT Server, so this PR is more or less only the groundwork to make it possible to configure VCSs.

fviernau commented:
> Yes., that part is clear, my concern is that it is not obvious form the configuration file that these settings will only be applied in certain situations, namely when Git is used via the download command of the CLI, but not when Git is used by the scanner. This should at least be documented very clearly to avoid confusion.

If this is really only for the downloader, would this mean often re-configuration depending on what project you clone. Was it considered to make it CLI option?

fviernau commented:
> Would "provenance resolution" mean that the Analyzer has scanned for dependency management files?

Provenance resolution is done as part of the "scanner" stage.

wkl3nk commented:
@mnonnenmacher I added comments in the _reference.yml_ to express that the functionality is limited to the Downloader tools only in command line use (scripted use), and has (out of the box) no effect when the Downloader is used in programmatical way. I added hints in the commit message.

mnonnenmacher commented:
@wkl3nk We discussed this PR in the [Kotlin developer meeting](https://github.com/oss-review-toolkit/ort/wiki/Kotlin-Developer-Meeting) today and came to the conclusion that there are still too many open questions with regards to the approach that we need to clarify in a dedicated meeting. I will contact you directly to schedule it.

wkl3nk commented:
Some ideas were discarded, others were implemented in different ways. Work continues in #9697 and #9721.
